### PR TITLE
feat: Mobile Vitals - Native App Start

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Send native app start with the first transaction.

Bildschirmfoto 2022-02-10 um 14 57 13

Bildschirmfoto 2022-02-10 um 14 57 06

## :bulb: Motivation and Context

This PR aims to bring together all the pieces for sending the native app start and duration with the first transaction.

- The integration `MobileVitalsIntegration` is responsible for enriching the first transaction that is being sent with the native app start data.
- This PR introduces `SentryNativeWrapper`, which is responsible to handle all native SDK calls in a type-safe way. Other calls used throughout the SDK can be moved here over time.
- The `SentryTransaction` is extended with an optional list of `SentryMeasurement` objects.
- We set the end of the app startup with `SchedulerBinding.addPostFrameCallback()`. This default behaviour can be disabled and users can provide the app startup end themselves by calling `SentryFlutter.setAppStartFinish`.

## :green_heart: How did you test it?

- Added unit tests.
- Tested with example application.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

- This PR adds a static reference to `SentryFlutterOptions`, so that users can set the `appStartFinish` timestamp. We need to discuss if this is the appropriate way to do this.

### CODE REVIEW

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
import io.sentry.android.core.AppStartState
```

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
    val isColdStart = AppStartState.getInstance().isColdStart()
```

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
      result.error("1", "App start won't be sent due to missing appStartTime", null)
```

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
      result.error("1", "App start won't be sent due to missing isColdStart", null)
```

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: We could control that within the Dart side, so we don't even call the message channel in case we've sent the App start.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:141]: Maybe we should return `result.success` outside of the `else` branch, otherwise the function never returns, an option is to return an empty Map if necessary.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: Does it mean that the App start on Flutter only works if you have set the `SentryNavigatorObserver`?
We should offer a public API to let people call it manually in case they are not using the `SentryNavigatorObserver`.
e.g. `SentryFlutter.setAppStartFinish(DateTime time = DateTimow.now()` or something similar.

**Comment by ueman** [comment]
Awesome! I'm really looking forward to this.

I'll have a question, though, why does this depend on the `NavigationObserver`? Why not just add a listener for the first [`SchedulerBinding.addPostFrameCallback()`](https://api.flutter.dev/flutter/scheduler/SchedulerBinding/addPostFrameCallback.html)? Then it would work even without the `NavigationObserver`.

Also, the `/` route isn't necessarily present or the first route. The new flutter skeleton app (`flutter create -t skeleton new_flutter_template`) doesn't have it, iirc.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: Another alternative is given by @ueman and its even better, https://github.com/getsentry/sentry-dart/pull/749#issuecomment-1034946909

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (denrase) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:141]: The closures on `result` are called in all possible code paths though?

**Comment by denrase** [comment]
@ueman Thanks for your input! 

Actually these are things i wanted to discuss with you folks, as I was not sure where to put the finish of the app start tbh. It's living in the `SentryNavigator` for now, as we are creating the auto transactions there.

You can pretty much see this PR as a first draft that has all the pieces we need to have the app start, but it's not final where those pieces would live.

Maybe it's better to move this into an integration and somehow access the (first) transaction there? WDYT @marandaneto?

**Comment by denrase** [comment]
I moved everything into an integration, now we are independent of the `SentryNavigatorObserver`. Thanks for your feedback so far!

File: flutter/lib/src/sentry_flutter.dart
Comment (ueman) [inline flutter/lib/src/sentry_flutter.dart]: This works not only on Android/iOS/macOS but also on web, Windows and Linux?

File: flutter/lib/src/sentry_flutter_options.dart
Comment (ueman) [inline flutter/lib/src/sentry_flutter_options.dart]: I kinda believe that `true` isn't a good default value because you definitely get a wrong measurement if a user of the library calls init after the application has already been started. It would affect Sentry Mobile as well.

Maybe one can automatically detect such thing and work around it?

File: flutter/lib/src/sentry_flutter.dart
Comment (denrase) [reply flutter/lib/src/sentry_flutter.dart]: This is only available for the native iOS/Android. Will fix this.

File: flutter/lib/src/sentry_flutter_options.dart
Comment (denrase) [reply flutter/lib/src/sentry_flutter_options.dart]: I imagined two cases:

- User starts sentry right at the beginning and everything works out of the box.
- Users can opt out when configuring options and call the method optionally.

Both rely on users starting the SDK upon app start, which we for example don't do with the sentry mobile app. I'm not sure if we can detect this.

The other option would be to reverse this and make it as an opt-in option. Then we won't have incorrect data for this case.

**Review by brustolin — APPROVED** [review]
LGTM

File: flutter/lib/src/sentry_flutter_options.dart
Comment (ueman) [reply flutter/lib/src/sentry_flutter_options.dart]: We could use https://api.flutter.dev/flutter/dart-ui/PlatformDispatcher/frameData.html which has the `frameNumber` as an indicator. If that's not at 0 (or whatever the start count is) you can automatically opt out and print a warning.

The drawback: It's only available after, I think, Flutter 2.5. So you would have to do something like: 
```dart
try {
  final frameNumber = (WidgetsFlutterBinding.instance!.window as dynamic).frameData.frameNumber as int;
} on NoSuchMethodError catch (_) {
  // ...
}
```

Whether that's actually worth it, is up to you.

Comment (denrase) [reply flutter/lib/src/sentry_flutter_options.dart]: Ok, that would be a possibility, but might also lead to confusion to users. There could be edge cases where users would want auto, but would get cut off? Difficult to decide tbh, maybe @marandaneto can help us out here?

Comment (denrase) [reply flutter/lib/src/sentry_flutter_options.dart]: I'll check how it is done on the other SDKs

File: dart/lib/src/protocol/sentry_transaction.dart
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_transaction.dart:84]: Can we use a `map` instead? Similar to `SentryEvent#toJson` as well

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:141]: I'm not sure, currently, all the methods are returning `result.success("")` so it's just a hunch.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: Is it needed for some reason? we should be able to track `AppStartState` even without that.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]:
```suggestion
                "isColdStart": isColdStart
```

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: Is this going to error the App? We should just log out instead of throwing an error.

File: flutter/lib/src/default_integrations.dart
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart]: Maybe we should create a new file, the default_integrations are getting bigger, wdyt?

File: flutter/lib/src/default_integrations.dart
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart]: Maybe this should go with a try/finally

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: Is this going to stay in memory forever? or does it set to null after finished and sent?
Otherwise, we could set it to null ourselves once we know it's done.

File: flutter/lib/src/sentry_flutter_options.dart
Comment (marandaneto) [reply flutter/lib/src/sentry_flutter_options.dart]: For now, we can document that people can turn it off the feature if initing the SDK too late, most probably, and later, they can add their own measurements to their own transactions anyway.
They could already do it via the `SentryMeasurement` class, so we unblock this PR, but we have to document this behavior.
TLDR: I'd rather not have a boolean called `autoAppStartFinish` but rather a flag called `autoAppStart` which enable/disable the whole feature.

File: flutter/lib/src/sentry_flutter_options.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_flutter_options.dart]: This should not be in the options, but rather a dedicated class that holds the state of this feature or `_NativeAppStartEventProcessor`

File: flutter/lib/src/sentry_flutter.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_flutter.dart]: Similar to https://github.com/getsentry/sentry-dart/pull/749/files#r816044870
Seems like a dedicated class would do better

File: flutter/lib/src/sentry_flutter.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_flutter.dart]: When addressing https://github.com/getsentry/sentry-dart/pull/749/files#r816048252 and friends, we'd not need to do this.

File: flutter/lib/src/default_integrations.dart
Comment (marandaneto) [reply flutter/lib/src/default_integrations.dart]: Is the intention to implement the slow and frozen frames in the same integration? if so, then the name would make sense, otherwise, I'd rather call e.g. AppStartIntegration

File: flutter/lib/src/default_integrations.dart
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart]: I'd rather get the `SchedulerBinding` via ctor with default value of `SchedulerBinding.instance`.
The idea is also to log out if `SchedulerBinding` is null, so users know why the feature isn't working.
It can also be used for testing/mocking.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: This is not needed after https://github.com/getsentry/sentry-dart/pull/777
We can also remove https://github.com/getsentry/sentry-dart/pull/749/files#r815995037 if not needed.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (denrase) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: This will return the error to the caller on the Flutter side. We then log it in the `NativeWrapper` class and return `null`.

File: flutter/lib/src/default_integrations.dart
Comment (denrase) [reply flutter/lib/src/default_integrations.dart]: - Will move it out.
- Yes, the intention, for now, is to also do the frames here.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: Usually, errors are more expensive than returning "nothing", what do you think?

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (denrase) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: It's unused, will remove.

File: CHANGELOG.md
Comment (marandaneto) [inline CHANGELOG.md]: Should go under `Unreleased`

File: flutter/lib/src/default_integrations.dart
Comment (denrase) [reply flutter/lib/src/default_integrations.dart]: This method does not throw, as the possible error sources (json/native bridge) are handled internally and logged. Or do you mean something else?

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (denrase) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: Works without it with the current main branch. 👍

File: flutter/lib/src/default_integrations.dart
Comment (denrase) [reply flutter/lib/src/default_integrations.dart]: Default value did not work as it is expecting `const` which didn't work here, but i moved it to the ctor so we can inject 👍

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (denrase) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: We do return error on all other code paths, so i'd rather be consistent and also do it or change it in all instances. Do you have a feeling how large the impact of returning the error might be?

File: dart/lib/src/protocol/sentry_transaction.dart
Comment (denrase) [reply dart/lib/src/protocol/sentry_transaction.dart:84]: I'm not sure what you are referring to, could you hive me a hint? In `SentryEvent` we have list/array in the JSON, which we do not have here, it's a map of maps (Data Structure).

Comment (marandaneto) [reply dart/lib/src/protocol/sentry_transaction.dart:84]: Sure, it's just a nit tho, feel free to ignore it, I guess I meant to use the extension methods like `.map` and `.where` as the other methods do, but indeed, the others are a List as well.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: I don't know the impact, it's just a hunch.
The difference is, we return a `FlutterError` when the SDK is unable to do its work, e.g. `captureEnvelope` or `initNativeSdk`.
The `appStart` feature is just a measurement, so the SDK is still running and working, hence just logging would be fine, it's rather a semantical comment than a bug.

File: flutter/lib/src/default_integrations.dart
Comment (marandaneto) [reply flutter/lib/src/default_integrations.dart]: Makes sense, `fetchNativeAppStart` already does a try catch internally.

File: flutter/lib/src/sentry_flutter_options.dart
Comment (denrase) [reply flutter/lib/src/sentry_flutter_options.dart]: I have introduced `SentryNativeState` to hold the `appStartEnd` timestamp. It is similar to what we have on Android with `AppStartState`, only more general. I opted to use a static property in 'SentryFlutter' to hold the instance, similar to the channel, instead of making a singleton like on Android.

Not to sure about the naming, as its only one value for now. We don't really have a place, except for options, to hold state in `SentryFlutter`. Can't add it to a Hub, as this data is not available in the dart layer. And we can't access the integrations from the flutter side without accessing dart internal API AFAIK. So we cant just use an `Integration` or `Processor` to hold this state.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (denrase) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: Can we log in the integrations? I'm not sure we can get hold of the logger instance, not seeing it beeing used?

Comment (ueman) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: You can pass options to the integrations and the options have the logger. Quite a few integrations access the options.

Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: The Swift plugin logs out using the `print` method since it's just for debugging only anyway.
You can also use this approach, or as @ueman said.

File: flutter/lib/src/sentry_flutter_options.dart
Comment (marandaneto) [reply flutter/lib/src/sentry_flutter_options.dart]: I believe it should be its own class like on Android, iOS also have dedicated classes to hold the state of some features (e.g. App start and Slow/Frozen frames).
I'd rather use it as a singleton than holding a reference in the SentryFlutter static class.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (denrase) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:141]: We now return `success(null)` and `success(nil)` and just print a waring if something went wrong. Returning an empty string would result in another error, as the deserializer would fail.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (denrase) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: @ueman Sorry, i confused integrations with the native plugins here. I meant the latter and am now logging directly in the plugin if there was an issue.

Bildschirmfoto 2022-03-09 um 13 49 31

File: flutter/lib/src/sentry_native_state.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_native_state.dart]: We could rather do a private ctor, with a public Factory, that always returns the same instance, See e.g. `HubAdapter`, NoOpHub, etc...
So `SentryNativeState()` always return the same instance instead of `SentryNativeState.instance` with a static method.

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (marandaneto) [inline flutter/lib/src/integrations/mobile_vitals_integration.dart]:
```suggestion
	final measurements = event.measurements ?? [];
        measurements.add(nativeAppStart.toMeasurement(appStartEnd));
        return event.copyWith(measurements: measurements);
```

You're replacing every `measurements` here already in place, we'd like to append to the list.

File: flutter/lib/src/sentry_flutter.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_flutter.dart:145]: We have to fix the docs here

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (marandaneto) [inline flutter/lib/src/integrations/mobile_vitals_integration.dart]: We could move to a new file

**Review by marandaneto — APPROVED** [review]

File: flutter/lib/src/sentry_flutter_options.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_flutter_options.dart:169]: on React Native this is called `enableAutoPerformanceTracking` and it's meant for both features, this one and Slow/Frozen frames, wdyt?
If this is only about setting the `setAppStartEnd` automatically, we rather keep it.
I'd still offer an option called `enableAutoPerformanceTracking` to enable or disable the whole feature.

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (marandaneto) [inline flutter/lib/src/integrations/mobile_vitals_integration.dart]: Not sure if it makes a difference, but the App start (Start) is UTC, so maybe `DateTime.now().toUtc()`?

File: flutter/lib/src/sentry_flutter_options.dart
Comment (denrase) [reply flutter/lib/src/sentry_flutter_options.dart:169]: If we combine the two, users won't be able to set values manually, so we won't enable the native side collection if `enableAutoPerformanceTracking` is false.

Comment (denrase) [reply flutter/lib/src/sentry_flutter_options.dart:169]: Introduced this on the https://github.com/getsentry/sentry-dart/pull/772 PR.

**Comment by denrase** [comment]
@marandaneto I created a separate branch to merge this into like you suggested.

File: CHANGELOG.md
Comment (marandaneto) [inline CHANGELOG.md]: Should go under Unreleased

File: flutter/lib/src/event_processor/native_app_start_event_processor.dart
Comment (marandaneto) [inline flutter/lib/src/event_processor/native_app_start_event_processor.dart:32]: I left a comment about this in the other PR.

**Comment by marandaneto** [comment]
@denrase It's green so feel free to merge into the feature branch PR :)

**Pull request merged**
